### PR TITLE
Refactor memory management in aster-frame

### PR DESCRIPTION
**Pull request by junyang-zh**
~~There's still implementations to be done. But the upper level API of the page table is can be shown as an RFC.~~

~~Core implementation completed!~~

~~2024/03/30: most jobs complete~~

~~2024/04/18 The scope of this PR is expanded.~~

2024/04/22 The scope of this PR is shrank. Frame metadata will not be introduced in this PR.

- [x] A new page table that is thread-safe, performant, huge page aware and with multi-arch abstractions.
- [x] Introduce custom bits in the PTE abstraction to support TDX.
- [x] Add unit tests.
~~- [x] Rename `VmSegment` to `PmSegment`, `VmFrame` to `PhysFrame`.~~
~~- [ ] Introduce `FrameMetaContainer` (i.e. `struct page`).~~
- [x] Remove `MemorySet` entirely. PT is suffice for implementing `VmSpace`.
~~- [ ] Implement a better page table utilizing the frame metadata.~~
~~- [ ] Implement `vmalloc` and the real `VmSegment`, IOMem should use it and cancell kpf handler.~~
~~- [ ] Optimize the case that parent need to copy even when owning the frame.~~

### CODE REVIEW

File: framework/aster-frame/src/vm/page_table/cursor.rs
Comment (lrh2000) [inline framework/aster-frame/src/vm/page_table/cursor.rs:21]: This is a very general comment about the design of the page table cursor.

Could you please have a look at the way how Rust's standard library [implements `CursorMut`](https://github.com/rust-lang/rust/blob/f2c80133931a5e5dc194359c0451bfaed579b48d/library/alloc/src/collections/btree/map.rs#L2857-L2869) for `BTreeMap`? I follow this idea to [implement `CursorMut`](https://github.com/asterinas/xarray/blob/72a4067a65e8f94cfc193f5f19ebc981c1de9de7/src/cursor.rs#L412-L429) for `XArray`. I think it also works for the page table cursor.

For the current design, I think the definition of `PtrRef` looks wired.
```rust
type PtfRef<E, C> = Arc<SpinLock<PageTableFrame<E, C>>>;
```
 - There is no point to have an `Arc<_>`, because you do not really want to share the page table.
 - There is no point to have a `SpinLock<_>`, because you always have exclusive access to the page table.

It looks like a workaround to make the Rust compiler happy, and yes, the Rust compiler will accept the current code, but I do think it's misuse of both `Arc<_>` and `SpinLock<_>`.

File: framework/aster-frame/src/vm/page_table/mod.rs
Comment (lrh2000) [inline framework/aster-frame/src/vm/page_table/mod.rs:76]: This is essentially a copy of page tables, but represented in Rust structures, right? I do not see too much value in why we need to maintain this, especially after we manage to avoid the abuse of internal `Arc<_>`s and `SpinLock<_>`s.

File: framework/aster-frame/src/arch/x86/iommu/second_stage.rs
Comment (lrh2000) [inline framework/aster-frame/src/arch/x86/iommu/second_stage.rs:74]: This problem might be outside the scope of this PR, but I'd like to raise it here.

The following definition of `PageTableEntry`, where a `PageTableEntry` directly references a entry in the page table, violates the Rust's memory model.
```rust
#[repr(C)]
pub struct PageTableEntry(usize);
```

 - By Rust's memory mode, whenever someone creates a `&PageTableEntry`, the entry must be immutable. However, the accessed bit and the dirty bit of the entry might be set by the CPU at any time.
 - Similarly, whenever someone creates a `&mut PageTableEntry`, they claim the exclusive access to the entry. However, the accessed bit and the dirty bit may be modified by the CPU at any time.

Generally, I would suggest that we should use only [`ptr::read()`](https://doc.rust-lang.org/std/ptr/fn.read.html) and [`ptr::write()`](https://doc.rust-lang.org/std/ptr/fn.write.html) to access the page table entries. It's okay to consider this issue out of scope, but it has a lot of implications for how you design and implement other page table types, such as the cursors.

**Review by lrh2000 — COMMENTED** [review]
A slight look at this PR and a few comments about the design.

I don't think I have time to look deeper at this moment.

File: framework/aster-frame/src/vm/page_table/mod.rs
Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/mod.rs:76]: Yes it is a copy of page tables. I don't really want to introduce it in the first place. However, we need a data structure to keep track of the page table frames, which would be allocated and freed dynamically. A PTE can keep the pointer to the frame but it cannot own the frame due to technical restrictions.

The rationale for `Arc` is that the kernel page tables are shared.

Comment (lrh2000) [reply framework/aster-frame/src/vm/page_table/mod.rs:76]: > The rationale for Arc is that the kernel page tables are shared.

Well, that's indeed a very strong argument for the `Arc<SpinLock<_>>` structure. I simply forgot about it thing when I had a quick look at this PR. Thanks for pointing that out.

But if this is the case, I think that you need to take all the parent locks when you operate on the leaf. Otherwise, race conditions will lead to inconsistent results, won't they?

**Comment by junyang-zh** [comment]
> Traditionally, global page table frames are kernel page table frames that are shared by all page tables; non-global page table frames are user page table frames that are never shared between page tables. Do we expect framework users to always follow this rule?

In this PR, the answer is yes. The user's root page table instance have no read or write permission to the kernel page tables.

But I am worrying about concurrent page fault handling from multiple user threads. If we need to get the exclusive access when handling a page fault the mapping operations cannot be parallelized. However I am not indeed sure about the support for such case.

**Comment by lrh2000** [comment]
> But I am worrying about concurrent page fault handling from multiple user threads. If we need to get the exclusive access when handling a page fault the mapping operations cannot be parallelized. However I am not indeed sure about the support for such case.

I agree with you on that point. But allowing concurrent changes makes the implementation much more complicated, which is my concern. Should we implement a correct, simple, but not so efficient approach first, or should we just accept this complexity in this PR?

If we decide to accept the complexity, to avoid race conditions (and deadlocks), I think we need a proper locking protocol. Have a look at common tree-based locking protocols, for example, search "Tree Based Protocol" on [this page](https://www.geeksforgeeks.org/graph-based-concurrency-control-protocol-in-dbms/).

> https://lwn.net/Articles/906852/

Per-VMA locking is another matter entirely. But you remind me to check the Linux kernel implementation for this. Have a look at [Split page table lock](https://www.kernel.org/doc/html/latest/mm/split_page_table_lock.html).

**Review by tatetian — COMMENTED** [review]
Here is my initial round of feedbacks.

**Comment by tatetian** [comment]
> Since the page table APIs reside in the Asterinas framework, I'd like to clarify a design goal:
> 
> * How do we expect framework users to organize the page table?

The Framework's users do not have access to page tables directly. They work on the abstractions of `VmSpace`. It suffices for them to know that the user virtual addresses use 0 as the most significant bit, while the kernel ones have 1.

> Traditionally, global page table frames are kernel page table frames that are shared by all page tables; non-global page table frames are user page table frames that are never shared between page tables. Do we expect framework users to always follow this rule?

The GLOBAL bit is an implementation detail. The Framework's users do not have to worry about it. Actually, GLOBAL bit is an optimization to avoid unnecessary TLB flush. Even without setting it, the kernel address space can still be shared by all user processes (after they enter the kernel mode).

>   * It's obvious that we can have direct and exclusive access to user page table frames, because they are never shared.
>   * By forbidding access to kernel page table frames from the root of a user page table (we shouldn't allow it, should we?), the only way to access the kernel page table frames is through the global variable `KERNEL_PAGE_TABLE`. This means that we can have direct and exclusive access to the kernel page table frames, too.

Agree. It looks that it is sufficient to have two big locks. One protects against the concurrent manipulation of the kernel portion of the page table; the other is per-process, protecting against the concurrent manipulation of the user portion.

File: framework/aster-frame/src/vm/page_table/properties.rs
Comment (tatetian) [inline framework/aster-frame/src/vm/page_table/properties.rs:54]: This is architectural dependent. So it should be moved to `arch/`.

File: framework/aster-frame/src/vm/space.rs
Comment (tatetian) [inline framework/aster-frame/src/vm/space.rs:266]: As we discussed, it should not be a permission bit.

**Comment by tatetian** [comment]
@junyang-zh Have the current implementation set the global bit for all kernel-space page table entries?

File: framework/aster-frame/src/vm/page_table/properties.rs
Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/properties.rs:50]: No it is not architectural dependent :>

**Comment by junyang-zh** [comment]
> @junyang-zh Have the current implementation set the global bit for all kernel-space page table entries?

yes

**Comment by junyang-zh** [comment]
> > 2024/04/18 The scope of this PR is expanded.
> 
> For what purpose? Doesn't the page table refactoring alone already result in a pretty complex PR with thousands of lines changed?

It was decided since introducing frame metadata we could make the page table even better (we can own frame handles without overhead).

But now I think the scope is... too big to manage.

If @tatetian agrees I want to make it two steps. limiting this PR to [60309c7](https://github.com/asterinas/asterinas/pull/713/commits/60309c7e57b5a58329aed66db0db4dd777d13c1a).

File: framework/aster-frame/src/vm/page_table/cursor.rs
Comment (lrh2000) [inline framework/aster-frame/src/vm/page_table/cursor.rs:26]: Well, what is the rationale for this? Is this an existing locking protocol or something similar to the Linux kernel implementation? I cannot understand it, so is there some other justification?

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/cursor.rs:26]: It seems that the documentation is malformed. What is implemented is: any read or write accesses to nodes require exclusive access on the entire path from the root to the node.

It is a doc of a previous implementation, I'll update it.

As for rationale for the current implementation, I just think it is true,  sadly no other supporting materials.

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/cursor.rs:26]:
```rust
/// Any read or write accesses to nodes require exclusive access on the
/// entire path from the root to the node. But cursor can be created without
/// holding the lock, and can release the lock after yeilding the current
/// slot while querying over the page table with a range. Simultaneous
/// reading or writing to the same range in the page table will not produce
/// consistent results, only validity is guaranteed.
```

Here's the updated doc. Since the cursor is an inner implementer of map/unmap/query operations and cursors are created in controlled contexts, I want to ensure consistency in higher abstraction levels (such as holding extra range locks in `map()`) since currently the implementation of cursor is sound.

Comment (lrh2000) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: > What is implemented is: any read or write accesses to nodes require exclusive access on the entire path from the root to the node.

What I mean is that I don't think it's right. It's not right even if we forbid "reading or writing to the same range in the page table".

For example, suppose the last level page table looks like this:
```
---------*-----------------------------------------------
         ^                                  ^
         |                                  |
     PTE x (mapped)                     PTE y (empty)
```

Now, the user program unmaps `PTE x` and maps `PTE y` simultaneously.
 1. First, we create a cursor for `PTE x` and another cursor `PTE y`.
 2. The second cursor unmaps `PTE y` first. Since the last level page table becomes empty, it gets released.
 3. However, a reference to the last level page table can still exist in the first cursor (in `stack: [Option<PtfRef<E, C>>; C::NR_LEVELS],`), which will cause problems. 

Note that, `acquire_locks` and `release_locks` have nothing to do with this problem. They acquire the locks, but do not ensure all the locked page tables are still alive.

What I still want to see is to implement a proper lock protocol. For example, I can suggest the followings:
 - `PageTableCursor::new` should accept an address range, which represents that the cursor will not (and cannot) touch page table entries that represent virtual addresses outside that range.
   - Then, `PageTableCursor::new` finds an intermediate page table (not necessarily the last-level or the top-level) which represents an address range that contains the whole specified address range. It requires all locks from the root page table to the intermediate page table, but then unlocks all locks excluding the one for the intermediate page table.
 - `PageTableCursor` then maintains the lock guards from one for the intermediate page table to the leaf. It shouldn't have `stack: [Option<PtfRef<E, C>>; C::NR_LEVELS],` anymore. Instead, it should only have `lock_guard: [Option<ArcSpinLockGuard<PageTableFrame<E, C>>>; C::NR_LEVELS],`, which starts from the one for the intermediate page table to the leaf page table.

For example, if we're going to map the address range shown below:
```
Top-level page tables             A
                                 /
                                B
                               / \
Last-level page tables        C   D
Last-level PTEs           ---**   **---
                             ^     ^
                             |     |
                             +--+--+
                                |
                Address range that we're going to map
```
 - When calling `PageTableCursor`, it will:
   - `lock(A)`, `lock(B)`, `unlock(A)`
   - `lock_guard = [ locked(B) ]`
 - When calling `PageTableCursor::map`, it will:
   - `lock(C)`, `lock_guard = [ locked(B), locked(C) ]`
   - Map some pages in `C`
   - `unlock(C)`, `lock_guard = [ locked(B) ]`
   - `lock(D)`, `lock_guard = [ locked(B), locked(D) ]`
   - Map some pages in `D`
   - `unlock(D)`, `lock_guard = [ locked(D) ]`

How do you like this idea? AFAIK, it should be deadlock-free and race-condition-free, even if we make `PageTableCursor` public and allow two cursors to operate on the same address range.

> Since the cursor is an inner implementer of map/unmap/query operations and cursors are created in controlled contexts, I want to ensure consistency in higher abstraction levels (such as holding extra range locks in `map()`) since currently the implementation of cursor is sound.

I don't think we hold any locks in the current implementation of `map()`, do we? Even worse, if we actually hold a lock in `map()`, additional locks in the cursor are pointless, right?

Comment (lrh2000) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: > How do you like this idea?

Indeed, I just realize one problem. With this approach, we cannot easily release empty page tables up to the root page table.
 - In the above example, even if `B` becomes empty, we cannot directly release it. It is because while holding the lock of `B`, we cannot relock `A`. Otherwise, it may cause deadlocks.

This can be fixed by adding another step. If the intermediate page table `B` becomes empty, we will first release its lock. Then, we try to acquire the locks from the root page table to the intermediate page table again (i.e., `lock(A)`, `lock(B)`) and recheck whether `B` is still empty. By doing so, we can easily release empty pages along the path.

I think it's acceptable and I cannot figure out a better way to resolve it.
 - An alternative is to hold all locks from the root page table (i.e., `A`), not the intermediate page table (i.e., `B`), to the leaf page table in the cursor. But it's pointless because it means that there is a big lock in the root page table and all other locks are pointless.
 - The root cause is that we cannot hold page tables (in `stack`) without holding their locks (in `stack_guard`). Otherwise, we'll have race conditions. So we should hold only page table lock guards `stack_guard` and remove `stack`. However, it means that we lost track to the unlocked pages without trying to relock them from the root page table again.

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: Marvelous. Indeed the current implementation could render another cursor invalid if a cursor unmaps a PTE, which is not as valid as I thought it could be.

Give me some time trying out the protocol you proposed. Thanks.

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: Well, here's my plan fixing this:
 - The cursor only hold lock guards, as you suggested;
 - When creating a cursor, a range is needed and the cursor will ensure exclusive ownership over that range. This is done by holding a lock of a page table containing that range, as you suggested;
 - However, when the cursor is created, it only drop the lock on the root node if applicable, allowing us for at most 512 ready cursors at the same time;
 - 2nd level page tables won't be freed by a cursor unless explicity recycled by another method. That introduces minimal wastage, but the wastage is desired for the kernel part(upper 256).

Comment (lrh2000) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: > * However, when the cursor is created, it only drop the lock on the root node if applicable, allowing us for at most 512 ready cursors at the same time;
> 
> * 2nd level page tables won't be freed by a cursor unless explicity recycled by another method. That introduces minimal wastage, but the wastage is desired for the kernel part(upper 256).

That does sound right.

The only potential concern is that it prevents concurrent access to the page table if both virtual addresses fall into the same entry of the second-level page table. So I guess you prefer this approach mainly for code simplicity?

Assume that the answer to the above question is yes. I'm not sure how well it does the job, since re-walking the page table to free the intermediate page table doesn't look very complicated either (and does not bring much runtime overhead since it only occurs if the intermediate page table becomes empty).

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: The problem with the rewalk is that you need to hold the reference to the root node just like what you did with the stack.

Maybe we can prove it correct even for the copy-on-write page table. As long as we mark cursors as RW cursors and Read only cursors.

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: Well I pondered upon it again. I think you are right. But it is a little bit trickier to avoid deadlocks: If a cursor has the lock of B and wants to recycle B, and another one tries to create a cursor mapping in the range of B, there would be a deadlock.

I haven't found a solution to it, since the pending to be created cursor will still be locked by another and the creator cannot tell if it is pending to be recycled in previous modifications to it.

Comment (lrh2000) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: > The problem with the rewalk is that you need to hold the reference to the root node just like what you did with the stack.

Well, the page table root is special. Because the root should never be deallocated (unless the whole page table is deallocated, of course) or replaced, should it?

If you do not want to keep a direct reference to the page table root, you can also reference the page table itself, i.e. `&'a PageTable`. This will make the cursor look even more correct (as usual, a cursor must have a lifetime parameter to prevent unexpected outlives).

By the way, I don't understand why `PageTableCursor` has an unused lifetime parameter `'a` in the current implementation. Why doesn't the compiler give a warning to make the CI fail?

```rust
pub(super) struct PageTableCursor<
    'a,
    M: PageTableMode,
    E: PageTableEntryTrait,
    C: PageTableConstsTrait,
> where
    [(); C::NR_ENTRIES_PER_FRAME]:,
    [(); C::NR_LEVELS]:,
{
    stack: [Option<PtfRef<E, C>>; C::NR_LEVELS],
    lock_guard: [Option<ArcSpinLockGuard<PageTableFrame<E, C>>>; C::NR_LEVELS],
    level: usize,
    va: Vaddr,
}
```

> Maybe we can prove it correct even for the copy-on-write page table. As long as we mark cursors as RW cursors and Read only cursors.

COW page tables look much more complex. Honestly speaking, I haven't given it much thought.

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: > Well I pondered upon it again. I think you are right. But it is a little bit trickier to avoid deadlocks: If a cursor has the lock of B and wants to recycle B, and another one tries to create a cursor mapping in the range of B, there would be a deadlock.

Ok, I know how. You need to release the lock of B before rewalking, and check the child counter again when reached B.

Comment (lrh2000) [reply framework/aster-frame/src/vm/page_table/cursor.rs]: > Well I pondered upon it again. I think you are right. But it is a little bit trickier to avoid deadlocks: If a cursor has the lock of B and wants to recycle B, and another one tries to create a cursor mapping in the range of B, there would be a deadlock.

I mean that we should release the lock, re-walk the page table from the root, and re-check whether `B` is still empty, as mentioned in https://github.com/asterinas/asterinas/pull/713#discussion_r1582009463:

> This can be fixed by adding another step. If the intermediate page table `B` becomes empty, we will first release its lock. Then, we try to acquire the locks from the root page table to the intermediate page table again (i.e., `lock(A)`, `lock(B)`) and recheck whether `B` is still empty. By doing so, we can easily release empty pages along the path.

I think doing that is okay because it happens _only_ when we drop the cursor and _only_ if the intermediate page table becomes empty in the first place. The overhead is acceptable. Meanwhile, it gives us opportunities to release not only `B` but also all `B`'s parents.

File: framework/aster-frame/src/vm/page_table/properties.rs
Comment (tatetian) [inline framework/aster-frame/src/vm/page_table/properties.rs:117]: This `MapOp` trait, the multiple `impl` blocks for `FnOnce<(MapInfo,)>`, and the helper functions (`cache_policy_op` and `perm_op`) look an unnecessarily complex solution to me.

I am particularly unconformable with the two functions of `cache_policy_op` and `perm_op`, which take a closure to return a new closure. This is too magical and implicit to my taste. The very design goal of mapping an old `MapProperty` to a new one in a concise way can be fulfilled in a more idiomatic and explicit way, without compromising the conciseness.

This can be simply done by changing the signature of a user function like `PageTable::protect` from the original form

```rust
    pub(crate) fn protect(
        &self,
        vaddr: &Range<Vaddr>,
        op: impl MapOp, // Fn(MapInfo) -> MapProperty
    ) -> Result<(), PageTableError> {
```

to the following form

```rust
    pub(crate) fn protect<F: Fn(&mut MapProperty)>(
        &self,
        vaddr: &Range<Vaddr>,
        fn: F,
    ) -> Result<(), PageTableError> {
```

Changing the signature does not seem to have any impact on the efficiency or correctness of the `protect` method.

But the caller side can be simplified. For example, an user using the original API looks as follows.

```rust
pt.protect(&prot, perm_op(|p| p - VmPerm::W))
```

Now with the new API, the user code is still simple enough, but does not need the magic `perm_op` utility function.

```rust
pt.protect(&prot, |old_prop| old_prop.perm -= VmPerm::W)
```

File: framework/aster-frame/src/vm/kspace.rs
Comment (tatetian) [inline framework/aster-frame/src/vm/kspace.rs:31]: The exact behaviors of this method is not clear to me as it lack of Rust doc. The function returns an `Option<_>`, which seems to suggest **only a valid virtual address will have its physical address returned**. But apparently, there exists some virtual addresses that fall within `LINEAR_MAPPING_BASE_VADDR..=kernel_loaded_offset()`, but are not backed by any main memory or MMIO regions.

I think this API is fundamentally flawed. Translating a virtual address through a direct mapping or a page table satisfies two different use cases. Using a single API to address two distinct needs with two different costs is not wise.

I checked all current usage of this function. There are three usage: two of them in `io_mem.rs`, one in `virtio_queue.rs`. The usage in `virtio_queue` will be soon removed (as the usage itself is flawed). The two in `io_mem.rs` can only rely on the direct mapping, not the page. So it can use another API that only translate virtual addresses that fall within the direct mapping.

This problem is not introduced by this PR. But fixing it can be easily done in this PR, preferably in a separate commit. I suggest marking this API deprecated and providing well-defined alternatives.

**Review by tatetian — COMMENTED** [review]
Since you are updating this PR actively, I decide to submit my feedbacks so far, before they get outdated. There will be more to come. Thanks.

**Comment by junyang-zh** [comment]
@tatetian Thanks for the valuable feed back! I'll start fixing these as soon as possible.

BTW, the latest commit implements:
 - the fine grained page table lock protocol;
 - lighter cursor, heavier frame ref (which is easier to refactor in the future without touching the cursor);
 - marked many of the methods safe since the cursor is created with control now. Even many of the methods of the cursor is safe now thanks to the frame ref abstraction.

File: framework/aster-frame/src/arch/x86/iommu/second_stage.rs
Comment (junyang-zh) [reply framework/aster-frame/src/arch/x86/iommu/second_stage.rs:22]: I don't know exactly. I just followed the previous implementation by @sdww0.

File: framework/aster-frame/src/vm/space.rs
Comment (junyang-zh) [reply framework/aster-frame/src/vm/space.rs]: Now I allow the cursor to jump to a given address after initialized. The problem is resolved since we can use the same cursor to query and map.

File: framework/aster-frame/src/vm/page_table/mod.rs
Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/mod.rs:250]: Originally in our code base this is uncacheable.

However I explored [all cr3 writes in Linux](https://elixir.bootlin.com/linux/latest/C/ident/native_write_cr3) I found them neither setting PWT nor PCD. So I changed it to `Writeback` in this PR.

File: framework/aster-frame/src/arch/x86/iommu/second_stage.rs
Comment (junyang-zh) [reply framework/aster-frame/src/arch/x86/iommu/second_stage.rs:20]: Seems that IOMMU has designs for supporting legacy 32-bit devices so 32 bit address space here is sound.

File: framework/aster-frame/src/arch/x86/boot/multiboot2/mod.rs
Comment (junyang-zh) [reply framework/aster-frame/src/arch/x86/boot/multiboot2/mod.rs:61]: Well this was a mistake here. The `base_addr` should ever be a plain physical address.

**Comment by junyang-zh** [comment]
Ok, fixed the first batch. The refactor of page property affects `aster-nix` as well.

CI is deemed to fail before #807 gets merged.

**Review by tatetian — COMMENTED** [review]
Here is my feedbacks for the updated version.

File: framework/aster-frame/src/vm/page_table/mod.rs
Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/mod.rs:276]: I am considering performance issues since it is faster not to check locks. The MMU can't afford concurrent modifications while walking either. It is crucial for the framework user to ensure not modifying the page table while other threads can read from the address, so this should be satisfied.

However, we seems to have already messed up with it, even if this is implemented with the cursor. Maybe we could fix it in the future.

Comment (junyang-zh) [reply framework/aster-frame/src/vm/page_table/mod.rs:280]: Fix: I've already checked that a PTE is always written after the physical address (of the child node/frame) points to initialized objects. So as long as MMU walk holds this holds too.

**Comment by tatetian** [comment]
I am all good with the new version. Waiting for the CI results to come out.

**Comment by lrh2000** [comment]
@junyang-zh Does this PR fix #681?

**Comment by junyang-zh** [comment]
> @junyang-zh Does this PR fix #681?

Short answer: no.

Long answer: we should let framework users use the cursor to access the page table, just like the usage case in `VmSpace::map`, which can use the cursor to hold the lock to ensure checking before modification.

I think this PR covers too much things now. We can refactor `VmSpace` in the future.

**Comment by lrh2000** [comment]
> > @junyang-zh Does this PR fix #681?
> 
> Short answer: no.
> 
> Long answer: we should let framework users use the cursor to access the page table, just like the usage case in `VmSpace::map`, which gets a cursor to hold the lock to ensure checking before modification.
> 
> I think this PR covers too much things now. We can refactor `VmSpace` in the future.

OK, thanks for the explanation. It's just a question to know whether we should keep the issue open. I don't mean that you should fix it here.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

File: framework/aster-frame/src/trap/handler.rs
Comment (lrh2000) [inline framework/aster-frame/src/trap/handler.rs]: I'm a little curious as to how this change was made. Why change it to share by default?

Pages shouldn't be shared unless explicitly requested, right?

Or just for consistency, the linear mapping built at boot is not shared (at least the last time I checked), right?

Comment (junyang-zh) [reply framework/aster-frame/src/trap/handler.rs]: Oh you have spotted a mistake here. Please open up an issue for it.

**Pull request merged**
